cmd/convertor/builder/builder.go

```diff
@@ -95,41 +97,167 @@ func NewOverlayBDBuilder(ctx context.Context, opt BuilderOptions) (Builder, erro
 			Transport: transport,
 		},
 	})
-	engineBase, err := getBuilderEngineBase(ctx, resolver, opt.Ref, opt.TargetRef)
+	fetcher, err := resolver.Fetcher(ctx, opt.Ref)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to new fetcher: %w", err)
 	}
-	engineBase.workDir = opt.WorkDir
-	engineBase.oci = opt.OCI
-	engineBase.mkfs = opt.Mkfs
-	engineBase.vsize = opt.Vsize
-	engineBase.db = opt.DB
 
-	refspec, err := reference.Parse(opt.Ref)
+	_, src, err := resolver.Resolve(ctx, opt.Ref)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to resolve: %w", err)
+	}
+	var manifestDescs []v1.Descriptor
+	switch src.MediaType {
+	case v1.MediaTypeImageManifest, images.MediaTypeDockerSchema2Manifest:
+		manifestDescs = append(manifestDescs, src)
+	case v1.MediaTypeImageIndex, images.MediaTypeDockerSchema2ManifestList:
+		rc, err := fetcher.Fetch(ctx, src)
+		if err != nil {
+			return fmt.Errorf("failed to fetch index: %w", err)
+		}
+		defer rc.Close()
+		b, err := io.ReadAll(rc)
+		if err != nil {
+			return fmt.Errorf("failed to read index: %w", err)
+		}
+		var index v1.Index
+		if err := json.Unmarshal(b, &index); err != nil {
+			return fmt.Errorf("failed to parse index: %w", err)
+		}
+		// (TODO) platform filter
+		manifestDescs = append(manifestDescs, index.Manifests...)
 	}
-	engineBase.host = refspec.Hostname()
-	engineBase.repository = strings.TrimPrefix(refspec.Locator, engineBase.host+"/")
-	engineBase.reserve = opt.Reserve
-	engineBase.noUpload = opt.NoUpload
-	engineBase.dumpManifest = opt.DumpManifest
-
-	var engine builderEngine
-	switch opt.Engine {
-	case Overlaybd:
-		engine = NewOverlayBDBuilderEngine(engineBase)
-	case TurboOCI:
-		engine = NewTurboOCIBuilderEngine(engineBase)
+	if len(manifestDescs) == 0 {
+		log.G(ctx).Warn("no matched manifest, do nothing")
+		return nil
 	}
-	return &overlaybdBuilder{
-		layers: len(engineBase.manifest.Layers),
-		engine: engine,
-		config: engineBase.config,
-	}, nil
+
+	var mu sync.Mutex
+	var targetDescs []v1.Descriptor
+	g, gctx := errgroup.WithContext(ctx)
+	for _, _mDesc := range manifestDescs {
```

Comment:
Following the introduced logic, we are basically creating an engine per manifest to convert each sub manifest. This introduces some concurrency concerns, especially because each conversion can potentially take up substantial storage space, we may want to add a limit to the number of manifests that can be converted at once and maybe make that configurable.

Reply:
add `--concurrency-limit`, default 4.